In [2]:
import pandas as pd
import numpy as np
import os

## Import “products.csv” data set from “Original Data” folder as df_prods and “orders_wrangled.csv” data set from “Prepared Data” folder as df_ords.

## Create path

In [3]:
path = r'C:\Users\inesh\Instacart Basket Analysis_01_02_2024'

Import “products.csv” as df_prods

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

Import orders_wrangled.csv as  df_ords

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col= 0) 

# (As the Orders_wrangled.csv date sheet was creating an unnamed index column, I used the index_col=0 argument)

In [6]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_from_last_order  
0                        8                   NaN  
1                        7                  15.0  
2                       12                  21.0  
3                        7                  29.0  
4                       15                  28.0  
...                    ...                   ...  
3421078                 18                  29.0  
3421079                 10                  30.0  
3421080                 12                  18.0  
3421081                 12                   7.0  
3421082                 14                  30.0  

[3421083 rows x 6 columns]

In [7]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_from_last_order'],
      dtype='object')

In [8]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_from_last_order  
count       3.421083e+06          3.214874e+06  
mean        1.345202e+01          1.111484e+01  
std         4.226088e+00          9.206737e+00  
min         0.000000e+00          0.000000e+00  
25%         1.000000e+01          4.000000e+00  
50%         1.300000e+01          7.000000e+00  
75%         1.600000e+01          1.500000e+01  
max         2.300000e+01          3.000000e+01

### Create test data frame

In [9]:
df_test = pd.DataFrame()

In [10]:
df_test.columns

RangeIndex(start=0, stop=0, step=1)

In [11]:
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

### Check dataframe for any mixed-type columns

In [13]:
for col in df_test.columns.tolist():
    weird= (df_test[[col]].map(type)!=df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird])>0:
        print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

### Finding missing values

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

### Create subset of data to show NaN values

In [16]:
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull()== False]

In [20]:
df_prods_clean.shape

(49677, 5)

### Finding Duplicates

In [21]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Remove Duplicates

In [23]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

## 4.5 Task

### Q2: Run the df.describe() function on your df_ords dataframe.

In [26]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_from_last_order  
count       3.421083e+06          3.214874e+06  
mean        1.345202e+01          1.111484e+01  
std         4.226088e+00          9.206737e+00  
min         0.000000e+00          0.000000e+00  
25%         1.000000e+01          4.000000e+00  
50%         1.300000e+01          7.000000e+00  
75%         1.600000e+01          1.500000e+01  
max         2.300000e+01          3.000000e+01

#####  Even though orders_day_of_week columns show 0 as the minimum value it is due to the the way that Data set labeled Sunday (0= Sunday). So, there are no anomalies in descriptive data of df_ords

#### Run the df.describe() function on  df_prods dataframe

In [36]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### *In the Price column the mean price is 9.994136. But and the standard deviation is 453.519686 and the maximum price is 99999, which is significantly higher than the other values. Therefore anomaly lies in the prices column. This can be an outlier or error.* 

### Q3: Check for mixed-type data in your df_ords dataframe

In [28]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].map(type)).any(axis = 1) # as I got an error with applymap() I used map()
  if len (df_ords[weird]) > 0:
    print (col)

##### There are no mixed data in any column

#### Q4 If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

##### *No mixed data-type columns found in df_ords*

#### Q5: Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

##### Finding missing values

In [29]:
df_ords.isnull().sum()

order_id                     0
user_id                      0
order_number                 0
orders_day_of_week           0
order_hour_of_day            0
days_from_last_order    206209
dtype: int64

##### There are 206209 missing values in the days_from_last_order column. Even though these appear as missing values, null values in the days_from_last_order column mean there is no gap between the last order & next order.  However, it is not appropriate to delete 206209 missing values

#### Q6: Address the missing values using an appropriate method

##### Create subset of data to show NaN values of days_from_last_order

In [30]:
df_ords_missing_value = df_ords[df_ords['days_from_last_order'].isnull()== True]

In [31]:
df_ords_missing_value

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_from_last_order  
0                        8                   NaN  
11                      11                   NaN  
26                      14                   NaN  
39                      11                   NaN  
45                      12                   NaN  
...                    ...                   ...  
3420930                 12                   NaN  
3420934                 18                   NaN  
3421002                 19                   NaN  
3421019                 15                   NaN  
3421069                 11                   NaN  

[206209 rows x 6 columns]

##### As mentioned above null values in the days_from_last_order column mean there is no gap between the last order & next order. Therefore after checking with data owners, I would like to update all these NaN values with 0

#### Q7: Run a check for duplicate values in your df_ords data

#####  Finding Duplicates

In [32]:
df_ords_duplicates = df_ords[df_ords.duplicated()]

In [33]:
df_ords_duplicates

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_from_last_order]
Index: []

##### There are no Duplicates in df_ords

### Q8 Address the duplicates using an appropriate method

##### As there are no duplicate, No action needed to remove duplicates.

### Q9: Export final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder

In [34]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'product_clean.csv'))

In [35]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_clean.csv'))